PayloadPkg/OsLoader/LoadImage.c

```diff
@@ -527,63 +463,105 @@ GetTraditionalLinux (
   return Status;
 }
 
-
 /**
   Load Image from boot media.
 
   This function will initialize OS boot device, and load image based on
   boot option, the loaded image info will be saved in  LoadedImage.
 
   @param[in]  BootOption     Current boot option
+  @param[in]  HwPartHandle   Hardware partition handle
+  @param[in]  FsHandle       FileSystem handle
   @param[out] LoadedImage    Loaded Image information.
 
   @retval  RETURN_SUCCESS    If image was loaded successfully
   @retval  Others            If image was not loaded.
 **/
 EFI_STATUS
-GetImageFromMedia (
-  IN  OS_BOOT_OPTION         *BootOption,
-  OUT LOADED_IMAGE           *LoadedImage
+EFIAPI
+LoadBootImage (
+  IN  OS_BOOT_OPTION  *OsBootOption,
+  IN  EFI_HANDLE       HwPartHandle,
+  IN  EFI_HANDLE       FsHandle,
+  OUT LOADED_IMAGE   **LoadedImagePtr
   )
 {
-  RETURN_STATUS              Status;
-  EFI_HANDLE                 FsHandle;
-  EFI_HANDLE                 HwPartHandle;
-  EFI_HOB_GUID_TYPE          *GuidHob;
-  EXT_BOOT_LOADER_VERSION    *VersionHob;
-
-  FsHandle     = NULL;
-  HwPartHandle = LoadedImage->HwPartHandle;
-  if (BootOption->FsType >= EnumFileSystemMax) {
-    Status = GetIasImageFromRawPartition (BootOption, LoadedImage);
-    return Status;
-  }
+  LOADED_IMAGE              *LoadedImage;
+  UINT8                     Index;
+  UINT8                     Count;
+  EFI_STATUS                Status;
+  EFI_HOB_GUID_TYPE         *GuidHob;
+  EXT_BOOT_LOADER_VERSION   *VersionHob;
+  BOOT_IMAGE                *BootImage;
+  UINT8                     BootFlags;
 
-  // Load image from file system.
-  DEBUG ((DEBUG_INFO, "Init File system\n"));
-  Status = InitFileSystem (BootOption->SwPart, EnumFileSystemTypeAuto, HwPartHandle, &FsHandle);
-  if (EFI_ERROR (Status)) {
-    DEBUG ((DEBUG_INFO, "No partitions found, Status = %r\n", Status));
-    return Status;
+  ASSERT (OsBootOption != NULL);
+
+  BootImage = OsBootOption->Image;
+  BootFlags = OsBootOption->BootFlags;
+
+  //
+  // Allocate/update LOADED_IMAGE according to Boot Flags
+  //
+  if (BootFlags & BOOT_FLAGS_EXTRA) {
+    for (Count = 0; Count < MAX_EXTRA_IMAGE_NUM; Count++) {
+      if (BootImage[LOAD_IMAGE_EXTRA0 + Count].LbaImage.Valid == 0) {
+        break;
+      }
+    }
+    LoadedImage = (LOADED_IMAGE *)AllocateZeroPool (sizeof (LOADED_IMAGE) * Count);
+    for (Index = 0; Index < Count; Index++) {
+      LoadedImage[Index].HwPartHandle  = HwPartHandle;
+      LoadedImage[Index].LoadImageType = LOAD_IMAGE_EXTRA0 + Index;
+    }
+  } else {
+    LoadedImage = (LOADED_IMAGE *)AllocateZeroPool (sizeof (LOADED_IMAGE));
+    LoadedImage->HwPartHandle = HwPartHandle;
+    if (BootFlags & BOOT_FLAGS_TRUSTY) {
+      LoadedImage->LoadImageType = LOAD_IMAGE_TRUSTY;
+    } else {
+      LoadedImage->LoadImageType = LOAD_IMAGE_NORMAL;
+    }

```

Comment:
Looks the logic is not same with original one.
By original logic, in one boot option there could be 7 images (misc image, normal image, trusty image, 4 extra images) loaded at same boot. could we still support it?

Reply:
It's not yet. Let me update logic to load all Boot Images in next commit.

Reply:
The same logic has been updated in the latest push.